File: algorithms/python/dynamic_programming/5_LongestPalindromicSubstring.py
class Solution:
    """Palindrome is a string that reads the same from left to right and right to left
    """
    def longestPalindrome(self, s: str) -> str:
        """Returns the longest palindrome
        
        Args:
            s (str): input string
            
        Returns:
            res (str): longest palindrome string
        """
        res = ""
        resLen = 0
        
        for i in range(len(s)):
            # odd length - initialise the left and right character pointer to the centre
            l, r = i, i
            while l >= 0 and r < len(s) and s[l] == s[r]:
                if (r - l + 1) > resLen:
                    res = s[l:r+1]
                    resLen = r - l + 1
                l -= 1
                r += 1
            
            # even length
            l, r = i, i + 1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                if (r - l + 1) > resLen:
                    res = s[l:r+1]
                    resLen = r - l + 1
                l -= 1
                r += 1
        
        return res

File: algorithms/golang/5_LongestPalindromicSubstring.go
package longestPalindrome

func longestPalindrome(s string) string {
	res := ""
	resLen, l, r := 0, 0, 0

	for i := 0; i < len(s); i++ {
		l, r = i, i
		// equivalent to a while loop
		for l >= 0 && r < len(s) && s[l] == s[r] {
			if (r - l + 1) > resLen {
				res = s[l : r+1]
				resLen = r - l + 1
			}
			r += 1
			l -= 1
		}

		l, r = i, i+1
		// equivalent to a while loop
		for l >= 0 && r < len(s) && s[l] == s[r] {
			if (r - l + 1) > resLen {
				res = s[l : r+1]
				resLen = r - l + 1
			}
			r += 1
			l -= 1
		}
	}
	return res
}
